### PR TITLE
(FMT): Improve code formatting: Line Breaks

### PR DESCRIPTION
**Pull request by jajakobyly**
This is a continuation of #302, which concentrates on formatting line breaks between code blocks.

Creative test cases are very, very welcome.

Most inspiration was taken from [Python formatter](https://upsource.jetbrains.com/idea-ce/file/idea-ce-d93ab79ac5306c9a9880e16438147eafe137f039/python/src/com/jetbrains/python/formatter/PythonFormattingModelBuilder.java), [Go formatter](https://upsource.jetbrains.com/go-lang-idea-plugin/file/ede4f339b4968989ea1a5bfcfefa0894874fef81/src/com/goide/formatter/GoFormattingModelBuilder.java) and [Kotlin formatter](https://github.com/JetBrains/kotlin/tree/master/idea/formatter/src/org/jetbrains/kotlin/idea/formatter) [(2)](https://github.com/JetBrains/kotlin/tree/master/idea/src/org/jetbrains/kotlin/idea/formatter).

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/formatter/RustBlock.kt [hunk 1]
```diff
@@ -21,6 +26,23 @@ class RustBlock(
...
+            }
+
+            if (psi1.isTopLevelItem && psi2.isTopLevelItem) {
+                return RustSpacing.lineBreak(
+                    keepLineBreaks = ctx.commonSettings.KEEP_LINE_BREAKS,
+                    keepBlankLines = ctx.commonSettings.KEEP_BLANK_LINES_IN_DECLARATIONS)
+            }
+        }
         return ctx.spacingBuilder.getSpacing(this, child1, child2)
     }
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/ide/formatter/RustBlock.kt]: I'm currently writing this the "classic" way - how it's implemented in Python, Go, Java, etc etc formatters. Though Kotlin formatter does this differently: it [greatly extends SpacingBuilder](https://github.com/JetBrains/kotlin/blob/master/idea/formatter/src/org/jetbrains/kotlin/idea/formatter/KotlinSpacingBuilder.kt) to empower creating complex spacing rules with a declarative way.

File: src/main/kotlin/org/rust/ide/formatter/RustBlock.kt [hunk 2]
```diff
@@ -21,6 +26,23 @@ class RustBlock(
...
+            val parent = psi1.parent
+
+            if (psi1 is RustOuterAttr && (psi2 is RustOuterAttr || parent is RustItem)) {
+                return RustSpacing.lineBreak()
+            }
+
+            if (psi1.isTopLevelItem && psi2.isTopLevelItem) {
+                return RustSpacing.lineBreak(
+                    keepLineBreaks = ctx.commonSettings.KEEP_LINE_BREAKS,
+                    keepBlankLines = ctx.commonSettings.KEEP_BLANK_LINES_IN_DECLARATIONS)
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/ide/formatter/RustBlock.kt]: This rule uses IDEA's default which states +/-: "keep blank lines, but maximally 2 consecutive"

File: src/main/kotlin/org/rust/ide/formatter/RustBlock.kt (hunk 1 above)
Comment (matklad) [inline src/main/kotlin/org/rust/ide/formatter/RustBlock.kt]: Judging by the end result, I am not sure that the Kotlin way is better :)

File: src/test/resources/org/rust/ide/intentions/fixtures/add_derive/add_derive_enum_after.rs [hunk 3]
```diff
@@ -1,3 +1,4 @@
-#[derive(<caret>)] enum Test {
+#[derive(<caret>)]
+enum Test {
```
Comment (matklad) [inline src/test/resources/org/rust/ide/intentions/fixtures/add_derive/add_derive_enum_after.rs]: :+1:

**Comment by matklad** [comment]
Looks good, but the tests are failing (so we have good tests, yay!)

**Comment by jajakobyly** [comment]
I prefer to open PRs early and receive code review as fast as possible ;) It's not everything I plan to create here

File: src/main/kotlin/org/rust/ide/formatter/RustBlock.kt [hunk 4]
```diff
@@ -58,33 +56,31 @@ class RustBlock(
...
-        private val BLOCK_START_TOKEN_SET = TokenSet.create(
-            PUB,
-            MOD,
-            STRUCT,
-            ENUM,
-            IMPL,
-            TRAIT,
-            MATCH
-        )
+private val BLOCK_START_TOKEN_SET = TokenSet.create(
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/formatter/RustBlock.kt]: It feels a bit inconsistent that we use either `TokenSet.create` or `ts`

File: src/main/kotlin/org/rust/ide/formatter/RustLanguageCodeStyleSettingsProvider.kt [hunk 5]
```diff
@@ -27,7 +27,10 @@ class RustLanguageCodeStyleSettingsProvider : LanguageCodeStyleSettingsProvider(
     override fun customizeSettings(consumer: CodeStyleSettingsCustomizable, settingsType: SettingsType) {
         when (settingsType) {
             BLANK_LINES_SETTINGS         -> {
-                // TODO Implement this.
+                consumer.showStandardOptions(
+                    "KEEP_LINE_BREAKS",
+                    "KEEP_BLANK_LINES_IN_DECLARATIONS",
+                    "KEEP_BLANK_LINES_IN_CODE")
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/formatter/RustLanguageCodeStyleSettingsProvider.kt]: Looks like there is really no way to avoid string literals here... Am I correct that it will not throw an error if you make a typo here?

File: src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt [hunk 6]
```diff
@@ -1,124 +1,163 @@
...
-        TYPE_WITH_BOUNDS_TYPE, FOR_IN_TYPE, WILDCARD_TYPE)
-    private val MACRO_ARGS = ts(MACRO_ARG, FORMAT_MACRO_ARGS, TRY_MACRO_ARGS)
-    private val PARAMS_LIKE = ts(PARAMETERS, VARIADIC_PARAMETERS)
-
-    inline fun SpacingBuilder.applyForEach(tokenSet: TokenSet,
-                                           block: SpacingBuilder.(IElementType) -> SpacingBuilder): SpacingBuilder {
-        var self = this
-        for (tt in tokenSet.types) {
-            self = block(this, tt)
+fun lineBreak(minLineFeeds: Int = 1,
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt]: can be private

File: src/main/kotlin/org/rust/ide/formatter/RustBlock.kt (hunk 4 above)
Comment (jajakobyly) [inline src/main/kotlin/org/rust/ide/formatter/RustBlock.kt]: I'm not sure, but it's possible that I will delete these guys when I will be working on indentation anyway.

File: src/main/kotlin/org/rust/ide/formatter/RustLanguageCodeStyleSettingsProvider.kt (hunk 5 above)
Comment (jajakobyly) [inline src/main/kotlin/org/rust/ide/formatter/RustLanguageCodeStyleSettingsProvider.kt]: Yep it just do not display given option in code style settings panel; so it's pretty easy to verify this anyway.

File: src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt [hunk 7]
```diff
@@ -1,124 +1,198 @@
...
+    get() = (this is RustItem || this is RustAttr) && this.parent is RustFile
+
+private val PsiElement.isStmtOrExpr: Boolean
+    get() = this is RustStmt || this is RustExpr
+
+private val ASTNode.hasLineBreakAfterInSameParent: Boolean
+    get() = treeNext != null && isWhiteSpaceWithLineBreak(TreeUtil.findFirstLeaf(treeNext))
+
+private val ASTNode.hasLineBreakBeforeInSameParent: Boolean
+    get() = treePrev != null && isWhiteSpaceWithLineBreak(TreeUtil.findLastLeaf(treePrev))
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt]: I'm not sure whether `TreeUtil` is available in non-idea ides; though these methods are almost copy-pasted from Python's formatter; so it has to be available at least in PyCharm.

File: src/main/kotlin/org/rust/lang/core/psi/util/RustPsiExtensions.kt [hunk 8]
```diff
@@ -22,16 +22,17 @@ inline fun <reified T : PsiElement> PsiElement.parentOfType(strict: Boolean = tr
 inline fun <reified T : PsiElement> PsiElement.childOfType(strict: Boolean = true): T? =
     PsiTreeUtil.findChildOfType(this, T::class.java, strict)
 
-fun PsiElement?.getNextNonPhantomSibling(): PsiElement? =
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/lang/core/psi/util/RustPsiExtensions.kt]: This method was dead code (and so the `PHANTOM_TOKEN_SET` used only in it), so I have allowed myself to just replace it with two sleeker alternatives.

File: src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt [hunk 9]
```diff
@@ -1,124 +1,198 @@
...
+            origPsi1.getPrevNonCommentSibling() ?: origPsi1
+        } else {
+            origPsi1
+        },
+        if (origPsi2 is PsiComment && origNode2.hasLineBreakBeforeInSameParent) {
+            origPsi2.getNextNonCommentSibling() ?: origPsi2
+        } else {
+            origPsi2
+        }
+    )
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt]: Ugh comments between formatting blocks are very burdensome. I hope the heuristics I have used here will work (almost) flawlessly.

Comment (matklad) [inline src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt]: Yep, comments are the hardest problem of every formatter

File: src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt [hunk 10]
```diff
@@ -1,124 +1,225 @@
...
+        val parentPsi = origPsi1.parent
+        val parentParentPsi = parentPsi.parent
+
+        // Here are rules which are comment-sensitive
+        when {
+            origPsi1 is RustOuterAttr && origPsi2 is PsiComment
+            -> return one()
+
+            origPsi1 is RustOuterAttr && (origPsi2 is RustOuterAttr || origPsi1.parent is RustItem)
+                || origPsi1 is PsiComment && (origPsi2 is RustOuterAttr || origPsi1.getPrevNonCommentSibling() is RustOuterAttr)
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt]: The hardest thing in programming is naming things. The second hardest it beautifully laying out complex logical expressions :|

Comment (matklad) [inline src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt]: The best way to format complex logical expressions which I am aware of is just not writing them. Extracting and naming a couple of sub expressions usually helps both formatting and clarity (although sometimes it interferes badly with order of evaluation). But I am not sure how to apply the technique in this particular case :)

Formatting [is hard](http://journal.stuffwithstuff.com/2015/09/08/the-hardest-program-ive-ever-written/)

**Comment by jajakobyly** [comment]
Ok, should work +/-. It doesn't play well with comments sometimes, but this formatter doesn't have to be the perfect one (rustfmt should be) and I can improve it later.

File: src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt [hunk 11]
```diff
@@ -1,124 +1,293 @@
...
+
+        //== empty parens
+        .between(LPAREN, RPAREN).spaceIf(false)
+        .between(LBRACK, RBRACK).spaceIf(false)
+        .between(LBRACE, RBRACE).spaceIf(false)
+        .betweenInside(OR, OR, LAMBDA_EXPR).spaceIf(false)
+
+        //== paren delimited lists
+        // withinPairInside does not accept TokenSet as parent node set :(
+        // and we cannot create our own, because RuleCondition stuff is private
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt]: You can write an extension method which uses cycle to simulate `TokenSet`, but it may not be a good idea.

**Comment by matklad** [comment]
Works good!

The code is hard to follow, but I don't think it can be made significantly simper. One trick to make is easier to read is to place Rust code examples in the comments for the trickiest parts.

**Comment by matklad** [comment]
Here are some examples of bad formatting: https://github.com/matklad/rustraytracer/commit/530c17b71dfddaae6345fa969af400a95a3199bf

They are mostly about bad indents, but there is also a spacing bug for `&`:

https://github.com/matklad/rustraytracer/commit/530c17b71dfddaae6345fa969af400a95a3199bf#diff-c14d5a69a77d82f4c014d3fd2fcb3361L40
https://github.com/matklad/rustraytracer/commit/530c17b71dfddaae6345fa969af400a95a3199bf#diff-73dec787512a58517803c1819029c229L112

**Comment by matklad** [comment]
Hm, can you reproduce https://youtrack.jetbrains.com/issue/KT-12239 ?

**Comment by jajakobyly** [comment]
> Hm, can you reproduce https://youtrack.jetbrains.com/issue/KT-12239 ?

Nope, works fine on my machine (I assume that this relates to Gradle build, so my spec shouldn't matter).
